File: developer-quarkus/src/main/java/de/dkutzer/buggy/developer/control/DeveloperRepository.java
package de.dkutzer.buggy.developer.control;

import com.google.common.collect.Lists;
import com.mongodb.client.FindIterable;
import com.mongodb.client.MongoClient;
import com.mongodb.client.MongoCollection;
import com.mongodb.client.model.Indexes;
import com.mongodb.client.model.UpdateOptions;
import de.dkutzer.buggy.developer.entity.Developer;
import io.quarkus.runtime.StartupEvent;
import org.bson.Document;
import org.bson.conversions.Bson;

import javax.enterprise.context.ApplicationScoped;
import javax.enterprise.event.Observes;
import javax.inject.Inject;
import java.util.*;

@ApplicationScoped
public class DeveloperRepository
{
    public static final String DEVELOPERS = "developers";
    public static final String BUGGY_DB = "buggy";

    @Inject
    MongoClient mongoClient;

    void onStart(@Observes StartupEvent ev) {

        getCollection().createIndex(Indexes.ascending("id"));
    }



    private UpdateOptions upsertOptions = new UpdateOptions().upsert(true);


    public Iterable<Developer> findAll() {

        final MongoCollection<Document> developersCol = getCollection();
        final FindIterable<Document> documents = developersCol.find();
        List<Developer> result = Lists.newArrayList();
        for(Document doc : documents){
            final Developer e = getDeveloperByDocument(doc);
            result.add(e);
        }
        return result;
    }

    private Developer getDeveloperByDocument(Document doc) {
        final Developer e = new Developer();
        e.setId(doc.getString("id"));
        e.setFirstName(doc.getString("firstName"));
        e.setLastName(doc.getString("lastName"));
        return e;
    }

    public Optional<Developer> findById(final String id){
        MongoCollection<Document> collection = getCollection();
        FindIterable<Document> iterable = collection.find(new Document("id", id));
        if(iterable.iterator().hasNext()){
            Document document = iterable.first();
            return Optional.of(getDeveloperByDocument(document));
        }
        return Optional.empty();
    }

    private MongoCollection<Document> getCollection() {
        return mongoClient.getDatabase(BUGGY_DB).getCollection(DEVELOPERS);
    }

    public void upsert(final Developer developer) {

        MongoCollection<Document> collection = getCollection();
        if (developer.getId() != null && !developer.getId().isEmpty()) {
            collection.updateOne(getFilterByDeveloper(developer), getDocumentByDeveloper(developer), upsertOptions);

        } else {
            collection.insertOne(getDocumentByDeveloper(developer));
        }
    }

    private Document getDocumentByDeveloper(Developer developer) {
        Map<String, Object> map = new HashMap<>();
        if (developer.getId() != null && !developer.getId().isEmpty()) {
            map.put("id", developer.getId());
        }
        map.put("firstName", developer.getFirstName());
        map.put("lastName", developer.getLastName());
        return new Document(map);
    }

    private Bson getFilterByDeveloper(Developer developer) {
        return new Document("id", developer.getId());
    }
    private Bson getFilterById(final String id) {
        return new Document("id", id);
    }

    public void delete(final String id){
        MongoCollection<Document> collection = getCollection();
        collection.deleteOne(getFilterById(id));
    }

    public boolean exists(String id) {

        return getCollection().countDocuments(getFilterById(id))!=0;
    }
}

File: developer-quarkus/src/main/kotlin/de/dkutzer/buggy/developer/control/DeveloperRepository.kt
package de.dkutzer.buggy.developer.control


import com.mongodb.client.MongoClient
import com.mongodb.client.MongoCollection
import com.mongodb.client.model.Filters
import com.mongodb.client.model.FindOneAndUpdateOptions
import com.mongodb.client.model.Indexes
import com.mongodb.client.model.ReplaceOptions
import de.dkutzer.buggy.developer.entity.Developer
import de.dkutzer.buggy.developer.entity.toEntity
import io.quarkus.runtime.StartupEvent
import org.bson.Document
import org.bson.conversions.Bson
import java.util.*
import javax.enterprise.context.ApplicationScoped
import javax.enterprise.event.Observes
import javax.enterprise.inject.Default
import javax.inject.Inject
import kotlin.collections.ArrayList

@ApplicationScoped
class DeveloperRepository {

    companion object {
        const val DEVELOPERS = "developers"
        const val BUGGY_DB = "buggy"
    }

    @Inject
    lateinit var mongoClient: MongoClient

    fun getCollection(): MongoCollection<Document> {
        return mongoClient.getDatabase(BUGGY_DB).getCollection(DEVELOPERS)
    }

    fun onStart(@Observes ev: StartupEvent?) {
        getCollection().createIndex(Indexes.ascending("id"))
    }


    private val findOneAndUpdateOptions = ReplaceOptions().upsert(true)


    fun findAll(): Iterable<Developer> {
        val developersCol = getCollection()
        val documents = developersCol.find()
        val result: MutableList<Developer> = ArrayList()
        documents.iterator().forEachRemaining {
            result.add(getDeveloperByDocument(it))
        }

        return result
    }

    private fun getDeveloperByDocument(doc: Document) = Developer(doc.getString("id"), doc.getString("firstName"), doc.getString("lastName"))

    fun findById(id: String?): Optional<Developer> {
        val collection = getCollection()
        val iterable = collection.find(Document("id", id))
        if (iterable.iterator().hasNext()) {
            val document = iterable.first()
            return Optional.of(getDeveloperByDocument(document))
        }
        return Optional.empty()
    }


    fun upsert(developer: Developer): Developer {
        val collection = getCollection()
        return if (developer.id != null && developer.id.isNotEmpty()) {
            collection.replaceOne(getFilterByDeveloper(developer), getDocumentByDeveloper(developer), findOneAndUpdateOptions)
            return developer
        } else {
            val document = getDocumentByDeveloper(developer)
            collection.insertOne(document)
            document.toEntity();
        }
    }

    private fun getDocumentByDeveloper(developer: Developer): Document {
        val map: MutableMap<String, Any> = HashMap()
        if (developer.id != null && developer.id.isNotEmpty()) {
            map["id"] = developer.id
        } else {
            map["id"] = UUID.randomUUID().toString()
        }
        map["firstName"] = developer.firstName
        map["lastName"] = developer.lastName
        return Document(map)
    }

    private fun getFilterByDeveloper(developer: Developer): Bson {
        return Filters.eq("id", developer.id)
    }

    private fun getFilterById(id: String): Bson {
        return Document("id", id)
    }

    fun delete(id: String) {
        val collection = getCollection()
        collection.deleteOne(getFilterById(id))
    }

    fun exists(id: String?): Boolean {
        return id != null && id.isNotEmpty() && getCollection().countDocuments(getFilterById(id)) != 0L
    }

    fun deleteAll() {

        getCollection().deleteMany(Document())
    }


}